pkg/results/report.go

```diff
@@ -0,0 +1,144 @@
+package results
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/golang/glog"
+
+	"github.com/openshift/ci-tools/pkg/api"
+)
+
+// Options holds the configuration options for connecting to the remote aggregation server
+type Options struct {
+	address  string
+	certFile string
+	keyFile  string
+	caFile   string
+}
+
+// Bind adds flags for the options
+func (o *Options) Bind(flag *flag.FlagSet) {
+	flag.StringVar(&o.address, "report-address", "", "Address of the aggregate reporting server.")
+	flag.StringVar(&o.certFile, "report-cert-file", "", "File holding the certificate for the aggregate reporting server.")
+	flag.StringVar(&o.keyFile, "report-key-file", "", "File holding the key for the aggregate reporting server.")
+	flag.StringVar(&o.caFile, "report-ca-file", "", "File holding the certificate authority for the aggregate reporting server.")
+}
+
+// Validate ensures that options are set correctly
+func (o *Options) Validate() error {
+	numSet := 0
+	for _, field := range []string{o.certFile, o.keyFile, o.caFile} {
+		if field != "" {
+			numSet = numSet + 1
+		}
+	}
+
+	if numSet != 0 && numSet != 3 {
+		return errors.New("--report-{cert|key|cacert}-file must be set together or not at all")
+	}
+	return nil
+}
+
+// Client returns an HTTP or HTTPs client, based on the options
+func (o *Options) Reporter(spec *api.JobSpec) (Reporter, error) {
+	r := &reporter{
+		spec:    spec,
+		address: o.address,
+		client:  &http.Client{},
+	}
+	if o.certFile == "" {
+		return r, nil
+	}
+
+	cert, err := tls.LoadX509KeyPair(o.certFile, o.keyFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not load client cert: %v", err)
+	}
+
+	caCert, err := ioutil.ReadFile(o.caFile)
+	if err != nil {
+		return nil, fmt.Errorf("could not load CA cert: %v", err)
+	}
+	caCertPool := x509.NewCertPool()
+	caCertPool.AppendCertsFromPEM(caCert)
+
+	tlsConfig := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      caCertPool,
+	}
+	tlsConfig.BuildNameToCertificate()
+	transport := &http.Transport{TLSClientConfig: tlsConfig}
+	r.client = &http.Client{Transport: transport}
+	return r, nil
+}
+
+// Request holds the data used to report a result to an aggregation server
+type Request struct {
+	// JobName is the name of the job for which a result is being reported
+	JobName string `json:"job_name"`
+	// Type is the type of job ("presubmit", "postsubmit", "periodic" or "batch")
+	Type string `json:"type"`
+	// State is "succeeded" or "failed"
+	State string `json:"state"`
+	// Reason is a colon-delimited list of reasons for failure
+	Reason string `json:"reason"`
+}
```

Comment:
Cluster field is missing

Reply:
Yeah, I thought about that one -- harder to do. Will attempt to do better now.